specs/agents/metadata.md

```diff
@@ -99,6 +99,109 @@ metadata is available.
+
+```sh
+curl -H "X-aws-ec2-metadata-token: $TOKEN" -v http://169.254.169.254/latest/meta-data
+```
+
+From the returned metadata, the following fields are useful
+
+| Cloud metadata field  | AWS Metadata field  |
+| --------------------  | ------------------- |
+| `account.id`          | `accountId`         |
+| `instance.id`         | `instanceId`        |
+| `availability_zone`   | `availabilityZone`  |
+| `machine.type`        | `instanceType`      |
+| `provider`            | aws                 |
+| `region`              | `region`            |
+
+#### GCP metadata
+
+Metadata about a GCP machine instance can be retrieved from the 
+metadata service, `http://metadata.google.internal`.
+
+An example with curl
+
+```sh
+curl -X GET "http://metadata.google.internal/computeMetadata/v1/?recursive=true" -H "Metadata-Flavor: Google"
+```
+
+From the returned metadata, the following fields are useful
+
+| Cloud metadata field  | AWS Metadata field  |
+| --------------------  | ------------------- |
+| `instance.id`         | `instance.id`       |
+| `instance.name`       | `instance.name`     |
+| `project.id`          | `project.numericProjectId` as a string |
+| `project.name`        | `project.projectId` |
+| `availability_zone`   | last part of `instance.zone`, split by `/`  |
+| `machine.type`        | last part of `instance.machineType`, split by `/` |
+| `provider`            | gcp                 |
+| `region`              | last part of `instance.zone`, split by `-`            |
+
+#### Azure metadata
+
+##### Azure VMs
+
+Metadata about an Azure VM can be retrieved from the internal metadata
+endpoint, `http://169.254.169.254`.
+
+An example with curl
+
+```sh
+curl -X GET "http://169.254.169.254/metadata/instance/compute?api-version=2019-08-15" -H "Metadata: true"
+```
+
+From the returned metadata, the following fields are useful
+
+| Cloud metadata field  | AWS Metadata field  |
+| --------------------  | ------------------- |
+| `account.id`          | `subscriptionId`    |
+| `instance.id`         | `vmId`              |
+| `instance.name`       | `name`              |
+| `project.name`        | `resourceGroupName` |
+| `availability_zone`   | `zone`              |
+| `machine.type`        | `vmSize`            |
+| `provider`            | azure               |
+| `region`              | `location`          |
+
+##### Azure App Services
+
+Azure App Services are a PaaS offering within Azure which doe not
+have access to the internal metadata endpoint. Metadata about
+an App Service can however be retrieved from environment variables
+
+
+| Cloud metadata field  | Environment variable |
+| --------------------  | ------------------- |
+| `account.id`          | first part of `WEBSITE_OWNER_NAME`, split by `+` |
+| `instance.id`         | `WEBSITE_INSTANCE_ID` |
+| `instance.name`       | `WEBSITE_SITE_NAME` |
+| `project.name`        | `WEBSITE_RESOURCE_GROUP` |
+| `provider`            | azure               |
```

Comment:
Is there value in capturing the cloud product, perhaps `cloud.product: azure-app-service` as well?  It appears we'd need a new [cloud ecs](https://www.elastic.co/guide/en/ecs/master/ecs-cloud.html) field.

Reply:
++ I think there could be. cloud fields look to be very much intended for VMs e.g. `availability_zone`, `machine_type`.

Reply:
With https://github.com/elastic/ecs/pull/1204 we will have an ECS field for this as of next release - `cloud.service.name`.  https://github.com/elastic/apm-server/issues/4625 is tracking addition to the intake for this.

Reply:
+1  for somehow capturing the fact that this set of metadata is being captured in an `azure-app-services` environment in order to make it distinct from data collected via the instance endpoints. I was thinking similar thoughts when I investigated this for node.js recently specifically around a theoretical problem of a user looking at two sets of `azure` data and wondering why different fields are collected from each.